app/src/org/commcare/preferences/FilePreferenceDialogFragmentCompat.java

```diff
@@ -0,0 +1,71 @@
+package org.commcare.preferences;
+
+import android.content.Intent;
+import android.net.Uri;
+import android.os.Bundle;
+import android.support.v7.preference.EditTextPreference;
+import android.support.v7.preference.EditTextPreferenceDialogFragmentCompat;
+import android.view.View;
+import android.widget.EditText;
+import android.widget.Toast;
+
+import org.apache.commons.io.FilenameUtils;
+import org.commcare.CommCareApplication;
+import org.commcare.dalvik.R;
+import org.commcare.utils.UriToFilePath;
+import org.javarosa.core.services.locale.Localization;
+
+import java.io.File;
+
+import static android.app.Activity.RESULT_OK;
+
+public class FilePreferenceDialogFragmentCompat extends EditTextPreferenceDialogFragmentCompat {
+
+    public static final int REQUEST_FILE = 1;
+    private EditText mEditText;
+
+    public static FilePreferenceDialogFragmentCompat newInstance(String key) {
+        final FilePreferenceDialogFragmentCompat fragment = new FilePreferenceDialogFragmentCompat();
+        final Bundle b = new Bundle(1);
+        b.putString(ARG_KEY, key);
+        fragment.setArguments(b);
+        return fragment;
+    }
+
+    @Override
+    protected void onBindDialogView(View view) {
+        super.onBindDialogView(view);
+        mEditText = (EditText)view.findViewById(android.R.id.edit);
+        view.findViewById(R.id.filefetch).setOnClickListener(v -> {
+            CommCarePreferences.startFileBrowser(FilePreferenceDialogFragmentCompat.this,
+                    REQUEST_FILE,
+                    ((FilePreference)getPreference()).getErrorDialogTitle());
+        });
+    }
+
+    @Override
+    public void onActivityResult(int requestCode, int resultCode, Intent intent) {
+        if (requestCode == REQUEST_FILE) {
+            if (resultCode == RESULT_OK && intent != null) {
+                Uri uri = intent.getData();
+                String filePath = UriToFilePath.getPathFromUri(CommCareApplication.instance(), uri);
+                String fileType = ((FilePreference)getPreference()).getFileType();
```

Comment:
Since you're checking for the proper file type as soon as it is selected here, I think all of the other places downstream from here where you've added checks for file type matching are unnecessary. Let me know if I'm missing something in that analysis